lib/backend/k8s/resources/workloadendpoint.go

```diff
@@ -165,60 +173,112 @@ func (c *WorkloadEndpointClient) Get(ctx context.Context, key model.Key, revisio
 	if !c.converter.IsValidCalicoWorkloadEndpoint(pod) {
 		return nil, cerrors.ErrorResourceDoesNotExist{Identifier: k}
 	}
-	return c.converter.PodToWorkloadEndpoint(pod)
+
+	kvps, err := c.converter.PodToWorkloadEndpoints(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find the WorkloadEndpoint that has a name matching the name in the given key
+	for _, kvp := range kvps {
+		wep := kvp.Value.(*apiv3.WorkloadEndpoint)
+		if wep.Name == key.(model.ResourceKey).Name {
+			return kvp, nil
+		}
+	}
+
+	return nil, kerrors.NewNotFound(apiv3.Resource("WorkloadEndpoint"), key.String())
 }
 
 func (c *WorkloadEndpointClient) List(ctx context.Context, list model.ListInterface, revision string) (*model.KVPairList, error) {
 	log.Debug("Received List request on WorkloadEndpoint type")
 	l := list.(model.ResourceListOptions)
 
-	// If a workload is provided, we can do an exact lookup of this
-	// workload endpoint.
+	// If a "Name" is provided, we may be able to get the exact WorkloadEndpoint or narrow the WorkloadEndpoints to a
+	// single Pod.
 	if l.Name != "" {
-		kvp, err := c.Get(ctx, model.ResourceKey{
-			Name:      l.Name,
-			Namespace: l.Namespace,
-			Kind:      l.Kind,
-		}, revision)
+		return c.listUsingName(ctx, l, revision)
+	}
+
+	return c.list(l, revision)
+}
+
+// listUsingName uses the name in the listOptions to retrieve the WorkloadEndpoints. The name, at the very least, must identify
+// a single Pod, otherwise an error will occur.
+func (c *WorkloadEndpointClient) listUsingName(ctx context.Context, listOptions model.ResourceListOptions, revision string) (*model.KVPairList, error) {
+	wepID, err := c.converter.ParseWorkloadEndpointName(listOptions.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if wepID.Pod == "" {
+		return nil, cerrors.ErrorResourceDoesNotExist{
+			Identifier: listOptions,
+			Err:        errors.New("malformed WorkloadEndpoint name - unable to determine Pod name"),
+		}
+	}
+
+	pod, err := c.clientSet.CoreV1().Pods(listOptions.Namespace).Get(wepID.Pod, metav1.GetOptions{ResourceVersion: revision})
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			return &model.KVPairList{
+				KVPairs:  []*model.KVPair{},
+				Revision: revision,
+			}, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	kvps, err := c.converter.PodToWorkloadEndpoints(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	// If Endpoint is available get the single WorkloadEndpoint
+	if wepID.Endpoint != "" {
+		wepName, err := wepID.CalculateWorkloadEndpointName(false)
 		if err != nil {
-			switch err.(type) {
-			// Return empty slice of KVPair if the object doesn't exist, return the error otherwise.
-			case cerrors.ErrorResourceDoesNotExist:
-				return &model.KVPairList{
-					KVPairs:  []*model.KVPair{},
-					Revision: revision,
-				}, nil
-			default:
-				return nil, err
+			return nil, err
+		}
+		// Find the WorkloadEndpoint that has a name matching the name in the given key
```

Comment:
What happens if we don't find a match here? I think we need to handle that.

e.g., if the wepID.Endpoint is an endpoint that doesn't actually exist.

Reply:
returning an empty list now